pytest_archon/rule.py

```diff
@@ -147,68 +147,100 @@ def check(
-        )
+        if not candidates:
+            log_failure(
+                f"NO CANDIDATES MATCHED. Match criteria: {match_criteria}, "
+                f"exclude_criteria: {exclude_criteria}",
+            )
+            return
 
         candidates = sorted(candidates)
 
-        if len(candidates) > 4:
-            candidates_to_show = candidates[:2] + ["..."] + candidates[-1:]
-        else:
-            candidates_to_show = candidates
-
-        print(f"rule {rule_name}: candidates are {candidates_to_show}")
-
-        for c in candidates:
-            imports = (
-                all_imports[c].get("direct", set())
-                if only_direct_imports
-                else all_imports[c].get("direct", set()) | all_imports[c].get("transitive", set())
-            )
+        for candidate in candidates:
+            import_map = {candidate: all_imports[candidate]} if only_direct_imports else all_imports
 
-            for constraint in self.ignored:
-                imports = {imp for imp in imports if not fnmatch(imp, constraint)}
-
-            for constraint in self.required:
-                matches = {imp for imp in imports if fnmatch(imp, constraint)}
-                check.is_true(
-                    matches,
+            for constraint in self._check_required_constraints(candidate, import_map):
+                log_failure(
                     _fmt_rule(
                         rule_name,
                         rule_comment,
-                        f"module '{c}' is missing REQUIRED imports matching pattern /{constraint}/",
+                        f"module '{candidate}' is missing REQUIRED imports matching pattern /{constraint}/",
                     ),
                 )
-            for constraint in self.forbidden:
-                matches = {imp for imp in imports if fnmatch(imp, constraint)}
-                check.is_false(
-                    matches,
+
+            for match, constraint, seen in self._check_forbidden_constraints(candidate, import_map):
+                log_failure(
                     _fmt_rule(
                         rule_name,
                         rule_comment,
-                        f"module '{c}' has FORBIDDEN imports:\n{matches} (matched by /{constraint}/)",
+                        f"module '{candidate}' has FORBIDDEN imports:\n{match} (matched by /{constraint}/), "
+                        f"through modules {' ↣ '.join(seen)}.",
                     ),
                 )
 
+    def _check_required_constraints(self, module: str, all_imports: ImportMap):
+        for constraint in self.required:
+            if not any(imp for imp in recurse_imports(module, all_imports) if fnmatch(imp, constraint)):
+                yield constraint
+
+    def _check_forbidden_constraints(
+        self,
+        module: str,
+        all_imports: ImportMap,
+    ):
+        seen = set()
+
+        def check(mod, path):
```

Comment:
I'm a bit in doubt: on one hand, we have better information on where it fails, but on the other hand we add more complexity: 

* we have the `check` and the `recurse_imports` function. Both do basically the same.
* we have to retrace the import graph for every constraint
* for every new check type, we have to add a recurse-type function
* recursion looks nice but is quite a heavy tool for python (at least I think python does not have any optimisations). If possible I use the iterative/for-loop version

But criticism without being constructive is not really helpful. What could we do alternatively and are the alternatives better?

I would perhaps distinguish between "test-mode" and "why-did-it-fail-mode". Test-mode should be fast, debugging why it failed probably not. The most common use case is probably a PR that adds depdencency, so you immediately know what went wrong. Only existing code-bases that suddenly are under the control of archon might be tricky. The only alternative I can think of is to supply a cli tool or debug flag for archon, that prints the dependency tree of the "candidate". The developer then could check by simple search & find what went wrong.

My argumentation would go this way: 99% of the cases we don't need to know why it failed. Now, in some cases we want to know, the "1%". Should we have this "1%" lead the implementation or go with the Perl motto: "Make Easy Things Easy and Hard Things Possible" while keeping the code lean.

Reply:
Also I would consider my points "complaining on a high level". Most likely this solution is fine and makes everybody happy.

Reply:
> I'm a bit in doubt: on one hand, we have better information on where it fails, but on the other hand we add more complexity:
> 
>     * we have the `check` and the `recurse_imports` function. Both do basically the same.

To some extend, yes. It may be worth extracting a higher level function out of it.
On the other hand: I tend to find recursive functions easier to read than functions with complex loops, and lots of state modifications.

>     * we have to retrace the import graph for every constraint

That's actually what we need for forbidden constraints to provide any context.

>     * for every new check type, we have to add a recurse-type function
> 
>     * recursion looks nice but is quite a heavy tool for python (at least I think python does not have any optimisations). If possible I use the iterative/for-loop version

Function calls used to be relatively expensive in Python 2. In Python 3 less so. The only "downside" to recursion is that you're limited to the stack depth. But if we'd reach that the code we test has some serious issues :grin: .

> But criticism without being constructive is not really helpful. What could we do alternatively and are the alternatives better?
> 
> I would perhaps distinguish between "test-mode" and "why-did-it-fail-mode". Test-mode should be fast, debugging why it failed probably not. The most common use case is probably a PR that adds a depdencency, so you immediately know what went wrong. Only existing code-bases that suddenly are under the control of archon might be tricky. The only alternative I can think of is to supply a cli tool or debug flag for archon, that prints the dependency tree of the "candidate". The developer then could check by simple search & find what went wrong.

The case I ran into was a bit more tricky: I updated a library in Gaphor, and that library imported a forbidden dependency in one of the modules where it previously did not. This was a bit of work to track, although I wrote all the code, both for Gaphor and the library.

> My argumentation would go this way: 99% of the cases we don't need to know why it failed. Now, in some cases, we want to know, the "1%". Should we have this "1%" lead the implementation or go with the Perl motto: "Make Easy Things Easy and Hard Things Possible" while keeping the code lean.

I kindly disagree on this one. We should provide as much context as possible on a failure. Those issues tend to happen because you import a module you think is "safe", but somewhere deep down a forbidden dependency is lurking, waiting to ruin your day.

Reply:
I updated `recurse_imports()` to return the path of imports. Now the forbidden imports could be reduced to a _one liner_.

Reply:
Nice!